[ONNX] Adds overload_name to Aten op (#69378)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73284
* #73283
* #73282
* #73281
* __->__ #73280

This PR adds a new attribute overload_name to the Aten node so that third party applications can implement calls to libtorch without using PyTorch source code.

This is necessary because torch's torch::jit::findOperatorFor(fullname) requires a full name, including operator and overload names.

ATen op was originally created for Caffe2, which leveraged the availability of the pytorch yaml files to create calls to the aten oeprators directly, not relying on torch::jit::findOperatorFor

The first part of the PR refactors all symbolics that create Aten ops, so that there is a single helper for this operator. Next all symbolics are updated to pass in the relevant overload name, if empty string is not applicable

Differential Revision: [D34625645](https://our.internmc.facebook.com/intern/diff/D34625645)

BowenBao committed:
[ONNX] Adds overload_name to Aten op (#69378)

This PR adds a new attribute overload_name to the Aten node so that third party applications can implement calls to libtorch without using PyTorch source code.

This is necessary because torch's torch::jit::findOperatorFor(fullname) requires a full name, including operator and overload names.

ATen op was originally created for Caffe2, which leveraged the availability of the pytorch yaml files to create calls to the aten oeprators directly, not relying on torch::jit::findOperatorFor

The first part of the PR refactors all symbolics that create Aten ops, so that there is a single helper for this operator. Next all symbolics are updated to pass in the relevant overload name, if empty string is not applicable

[ghstack-poisoned]

malfet commented on caffe2/contrib/aten/aten_op_template.h:
```diff
@@ -179,8 +179,9 @@ class ATenOp : public Operator<Context> {
     for (const auto i : c10::irange(operator_def.arg_size())) {
       auto & attr = operator_def.arg(i);
-      if(attr.name() == "operator" || attr.name() == "type" )
+      if(attr.name() == "operator" || attr.name() == "type" || attr.name() == "overload_name" ) {
```
Hmm, did you get an ack from someone who maintains caffe2 codebase? (Though not sure who owns it right now)
@ezyang do you know by any change?

thiagocrepaldi replied:
Not really, but this is back compatible. `overload_name` is an optional attribute which is ignored by Caffe2 with this exact condition :)

BowenBao committed:
Update on "[ONNX] Adds overload_name to Aten op (#69378)"


This PR adds a new attribute overload_name to the Aten node so that third party applications can implement calls to libtorch without using PyTorch source code.

This is necessary because torch's torch::jit::findOperatorFor(fullname) requires a full name, including operator and overload names.

ATen op was originally created for Caffe2, which leveraged the availability of the pytorch yaml files to create calls to the aten oeprators directly, not relying on torch::jit::findOperatorFor

The first part of the PR refactors all symbolics that create Aten ops, so that there is a single helper for this operator. Next all symbolics are updated to pass in the relevant overload name, if empty string is not applicable

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34625645).
